### PR TITLE
Generate WebP images

### PR DESCRIPTION
**Pull request by aslafy-z**
- [x] I've checked that this isn't a new swag opportunity proposal.
- [x] I've checked that this isn't a duplicate pull request.

- Generate some webp images using `gulp-responsive`.
- Use `<picture>` and `srcset` to reference it optionally (see https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images)

Ref #103

### CODE REVIEW

**Comment by vikaspotluri123** [comment]
So I haven't looked into webp support a bunch, but from what I remember reading a while ago, there's some server response involved with sending webp - I'm guessing Netlify can handle this because SmashingMag uses webp, but I'm curious if you have any insights 😇

**Comment by aslafy-z** [comment]
I thought `webp` whould work out of the box for most web servers but some might not set the `Content-Type` correctly because it's mime type is *still* not listed in the official [IANA mime list](https://www.iana.org/assignments/media-types/media-types.xhtml). Still, they all uses `image/webp` as the mime type for webp.

See https://bugzilla.mozilla.org/show_bug.cgi?id=600919#c139 and below.

Netlify seems to accept it:

```
$ curl -I https://deploy-preview-226--peaceful-chandrasekhar-efde8e.netlify.com/assets/swag-img/devrant-26d3018fd4.webp 
HTTP/2 200 
accept-ranges: bytes
content-length: 8678
content-type: image/webp           <===== YEAH
date: Wed, 12 Dec 2018 20:45:44 GMT
etag: "bc4160e1173cdf61f9e516bbd1a7638a-ssl"
feature-policy: accelerometer 'none'; camera 'none'; geolocation 'none'; gyroscope 'none'; magnetometer 'none'; microphone 'none'; payment 'none'; usb 'none'
referrer-policy: strict-origin-when-cross-origin
strict-transport-security: max-age=31536000; includeSubDomains; preload
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
x-nf-srv-version: 72b8ce8c0aacc79c990e81dd93dc325fb603aadd
x-robots-tag: noindex
x-xss-protection: 1; mode=block
age: 2
server: Netlify
cache-control: public max-age=63072000
x-nf-request-id: 2e347582-98d5-4571-ae26-84f5649895f9-4004685
```

**Comment by vikaspotluri123** [comment]
I've gotten some sleep, and I remember the issue - browsers that don't support webp!

**Comment by aslafy-z** [comment]
See https://github.com/swapagarwal/swag-for-dev/issues/103#issuecomment-428248731 

Just seen this article, from today! https://www.androidpolice.com/2018/12/12/firefox-beta-65-adds-support-for-webp-images/

**Comment by vikaspotluri123** [comment]
We have too many Safari users to ignore them 😕

**Comment by aslafy-z** [comment]
Have a look to https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images. The goal is not to drop jpeg but add webp support for browsers that supports it.

**Comment by vikaspotluri123** [comment]
Gotcha, didn't realize there was a plan to switch from `img` to `picture` (or something along those lines) 👍

**Comment by aslafy-z** [comment]
Still have to cleanup the webp filename, but it works!

File: gulpfile.js [hunk 1]
```diff
@@ -147,14 +154,15 @@ gulp.task('cachebust', cb => {
...
 			manifest = require(REV_PATH);
-
 			swagList.forEach(swag => {
-				const filename = `swag-img/${swag.image.split('/').pop()}`;
-				if (!manifest[filename]) {
-					console.warn('Unable to find image in manifest:', filename);
-					return;
-				}
-				swag.image = `/assets/${manifest[filename]}`;
+				Object.entries(swag.image).forEach(([extension, filename]) => {
```
Comment (vikaspotluri123) [inline gulpfile.js]: Wouldn't you be trying to destructure a string here?

File: gulpfile.js [hunk 2]
```diff
@@ -98,14 +98,29 @@ gulp.task('swag-img:download', () => {
 });
 
 gulp.task('swag-img:optimize', () => {
+	const imageSettings = {
```
Comment (vikaspotluri123) [inline gulpfile.js]: let's rename this to `jpegOutputConfig` for clarity

File: gulpfile.js [hunk 3]
```diff
@@ -98,14 +98,29 @@ gulp.task('swag-img:download', () => {
...
-			height: 300,
-			format: 'jpeg',
-			flatten: true
-		}], RESIZE_OPTS))
-		.pipe(gulp.dest('dist/assets/swag-img'));
+		.pipe(responsive([
+			// JPEG output
+			imageSettings,
+			// WebP output
+			Object.assign({
```
Comment (vikaspotluri123) [inline gulpfile.js]: Might be worth creating a variable for this outside of here for clarity

File: src/pug/includes/body.pug [hunk 4]
```diff
@@ -53,7 +53,11 @@ include ./fork.pug
                         each swagTag in swag.tags
                             span!=swagTag
                     div.flex.img-container
-                        img(src=swag.image alt=swag.name + " swag you can get")
+                        picture
```
Comment (vikaspotluri123) [inline src/pug/includes/body.pug]: 💃

File: gulpfile.js (hunk 1 above)
Comment (aslafy-z) [reply gulpfile.js]: Nope, see here: https://github.com/swapagarwal/swag-for-dev/pull/226/commits/1ae989bf18b8e03b9fa3fba59c28051cfa93dd29#diff-b9e12334e9eafd8341a6107dd98510c9R118. Just switched to `images` to be clearer.

File: src/pug/includes/body.pug (hunk 4 above)
Comment (aslafy-z) [reply src/pug/includes/body.pug]: Did I miss something?

Comment (vikaspotluri123) [reply src/pug/includes/body.pug]: Nope, it's a good thing!

File: gulpfile.js (hunk 1 above)
Comment (vikaspotluri123) [reply gulpfile.js]: whoops, I thought you were using `Object.values` - I think that would be better here?

Comment (aslafy-z) [reply gulpfile.js]: I reuse the extension from the object keys. If we want to use Object values I'll have to compute it from filename like so: 

```diff
diff --git a/gulpfile.js b/gulpfile.js
index af6b475..668cf4c 100644
--- a/gulpfile.js
+++ b/gulpfile.js
@@ -155,8 +155,9 @@ gulp.task('cachebust', cb => {
                        delete require.cache[require.resolve(REV_PATH)];
                        manifest = require(REV_PATH);
                        swagList.forEach(swag => {
-                               Object.entries(swag.images).forEach(([extension, filename]) => {
+                               Object.values(swag.images).forEach(filename => {
                                        filename = `swag-img/${filename.split('/').pop()}`;
+                                       const extension = filename.split('.').pop();
                                        if (!manifest[filename]) {
                                                console.warn('Unable to find image in manifest:', filename);
                                                return;
```

Comment (aslafy-z) [reply gulpfile.js]: Gonna keep `Object.entries` as it's fits best

File: gulpfile.js [hunk 5]
```diff
@@ -98,14 +98,30 @@ gulp.task('swag-img:download', () => {
...
+
 	return gulp.src('dist/assets/swag-img/*')
-		.pipe(responsive([{
-			name: '*',
-			height: 300,
-			format: 'jpeg',
-			flatten: true
-		}], RESIZE_OPTS))
-		.pipe(gulp.dest('dist/assets/swag-img'));
+		.pipe(responsive([
```
Comment (vikaspotluri123) [inline gulpfile.js]: This could probably fit in one line

File: gulpfile.js [hunk 6]
```diff
@@ -140,14 +155,15 @@ gulp.task('cachebust', cb => {
...
-					return;
-				}
-				swag.image = `/assets/${manifest[filename]}`;
+				Object.entries(swag.images).forEach(([extension, fileName]) => {
+					fileName = `swag-img/${fileName.split('/').pop()}`;
+					if (!manifest[fileName]) {
+						console.warn('Unable to find image in manifest:', fileName);
+						return;
+					}
+					swag.images[extension] = `/assets/${manifest[fileName]}`;
```
Comment (vikaspotluri123) [inline gulpfile.js]: didn't notice this earlier, gotcha!

File: gulpfile.js [hunk 7]
```diff
@@ -100,14 +100,27 @@ gulp.task('swag-img:download', () => {
...
-		.pipe(responsive([{
-			name: '*',
-			height: 300,
-			format: 'jpeg',
-			flatten: true
-		}], RESIZE_OPTS))
-		.pipe(gulp.dest('dist/assets/swag-img'));
+		.pipe(responsive([jpegOutputConfig, webpOutputConfig], RESIZE_OPTS))
+		.pipe(gulp.dest('dist/assets/swag-img'))
+		.on('end', () => {
```
Comment (vikaspotluri123) [inline gulpfile.js]: I'm pretty sure you need to use a callback if you add an `on('end')` listener because you run into race conditions

**Comment by plibither8** [comment]
Looks good :+1:, I think we can merge :)

**Comment by vikaspotluri123** [comment]
Thanks @aslafy-z! I was served webp in my browser ⚡ 🎉

**Pull request merged**